### DISCUSSION TITLE
REST Api / Flows: Update Item Bulk (unique values per Item)

### DISCUSSION THREAD

**Discussion by jakedowns** [directus/directus#16258] [Feature Requests]
It seems like the ability to pass an array of keys or ids, and a respective array of data, should be available as a means for updating multiple records.

I know bulk updating records is probably a rarer use case, but if you think of the example of a leaderboard collection, where, every time you add a new score, you want to re-calculate the ranking for all items, this use case currently requires crafting a mighty list of graphql mutations, or using REST / Flows, the use of a loop and singular update operations per row. which doesn't scale well when you have thousands of entries.

(Yes, i can just sort by the raw score values, but when leaderboards get long, it's helpful to be able to grab entries in a window AROUND a given Item/row and in that case, having a cached "rank" field (pre-sorted) is helpful rather than having to load ALL scores and then carve out the window you want to display)

**Comment by azrikahar** [comment]
This should actually be possible since `9.13.0` thanks to #14096! The PR itself revolves around Insights, but to highlight the second point in the PR:

> - Allow updating multiple items to multiple values in API
>      - Add `updateBatch` to services
>      - Add's `update_<collection>_batch` mutation in gql

### API

Here's a quick example:

https://user-images.githubusercontent.com/42867097/199862997-400e4390-f4d3-4381-913d-6256eb366202.mp4

That being said, it seem like this should be documented in the docs to highlight that this is indeed possible. Additional reference to `updateBatch` logic in `ItemsService`:

https://github.com/directus/directus/blob/7899f6a2679ea5f0ffcdb29e1d722be41ca0e5c3/api/src/services/items.ts#L439-L458

On that note, @licitdev this reminds me of the `updateBatch` you have added to `CollectionsService` in #16039, and it seems like it does not have the "preventing system cache from being cleared multiple times" optimization you did for `CollectionsService`. `ItemsService` should actually have the same optimization if I'm not mistaken right? 

### Flows

It does looks like Flows Update Data operation does not have this capability yet (as it was built to work "similarly" as other data operations), so that's definitely a potential enhancement to that specific operation 👍

**Reply by licitdev** [reply]
> `ItemsService` should actually have the same optimization if I'm not mistaken right?

@azrikahar Absolutely! Great spot on this optimisation~ 💯
Mind creating an issue to track this improvement?

**Reply by azrikahar** [reply]
Tracked in #16318! 👍

**Reply by renet** [reply]
> It does looks like Flows Update Data operation does not have this capability yet (as it was built to work "similarly" as other data operations), so that's definitely a potential enhancement to that specific operation 👍

Looking for this, as well. Is there an issue tracking this feature already, or should I open a new one?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by CMonjo** [comment]
Hi ! Sorry to up this discussion... I have the impression that it's impossible to do a bulk update with different values in graphql. 

Here is the directus doc : https://docs.directus.io/reference/items.html#update-multiple-items, in the GQL example, data is an unique value, which will be assigned to all ids, right?

If you have the slightest idea of how to bulk update with the possibility of setting a different data for each id 🙏

**Reply by pklenovic** [reply]
Hi @CMonjo. I managed to bulk update using flows. To be more specific using Webhook / Request URL flow operation.
So once you have your data in an array, pass it to body of Webhook / Request URL flow operation. Add proper headers, raw value:
`[
    {
        "header": "Content-Type",
        "value": "application/JSON"
    }
]`
Add access_token query parameter to you url. You can create it in your user profile. Just make sure once you created it that you save it locally and save changes made on your profile. Example `https://yourdomain.com/items/articles?access_token=your-access-token-value`
I also added picture of my flow and my webhook url. Hope this helps!

![Screenshot 2023-11-23 at 16 56 31](https://github.com/directus/directus/assets/123365244/1a8a92c4-4ccb-4c20-88e5-d8354e114d25)
![Screenshot 2023-11-23 at 16 48 36](https://github.com/directus/directus/assets/123365244/8e731d8e-9441-4776-94d4-a9ebba72bfbf)
